[FSDP2] Added 2D DCP save/load test

awgu opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #120952
* #121680
* #121816
* __->__ #121747

To prepare for FSDP2 + TP/SP in torchtrain, we should verify that we can resume training correctly with DCP save/load. For loading into a new model/optimizer instance, torchtrain uses lightweight `ModelWrapper` and `OptimizerWrapper`. In the added unit test, we use `get_optimizer_state_dict` directly to show the minimal requirement for correctness.

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @penguinwu @fegin @XilunWu @wanchaol @fduwjj @wz337 @tianyu-l @wconstab @yf225 @chauhang

Commit:
[FSDP2] Added 2D DCP save/load to same model test

[ghstack-poisoned]

Commit:
Update on "[FSDP2] Added 2D DCP save/load to same model test"




cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225 chauhang

[ghstack-poisoned]

awgu commented on torch/distributed/_composable/fsdp/_fsdp_param.py:
```diff
@@ -393,7 +393,13 @@ def _get_grad_inner_tensor(self, grad: torch.Tensor) -> torch.Tensor:
             if isinstance(grad, AsyncCollectiveTensor):
                 grad = grad.wait()
-            grad = cast(DTensor, grad)._local_tensor
+            assert isinstance(grad, DTensor), f"{type(grad)}"
```
Include this fix for partial gradients (e.g. for replicated `norm.weight`)

awgu commented on test/distributed/_composable/fsdp/test_fully_shard_training.py:
```diff
@@ -710,13 +712,16 @@ class TestFullyShard2DTraining(FSDPTest):
 
-    @skip_if_lt_x_gpu(2)
-    def test_train_parity_2d_mlp(self):
+    def init_global_mesh(self) -> DeviceMesh:
```
just factor this global mesh init into a method

awgu commented:
Convert to draft since I will look at loading into new model/optimizer instance and maybe include that in this PR too

Commit:
Update on "[FSDP2] Added 2D DCP save/load test"


To prepare for FSDP2 + TP/SP in torchtrain, we should verify that we can resume training correctly with DCP save/load.

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225 chauhang

[ghstack-poisoned]

Commit:
Update on "[FSDP2] Added 2D DCP save/load test"


To prepare for FSDP2 + TP/SP in torchtrain, we should verify that we can resume training correctly with DCP save/load. For loading into a new model/optimizer instance, torchtrain uses lightweight `ModelWrapper` and `OptimizerWrapper`. In the added unit test, we use `get_optimizer_state_dict` directly to show the minimal requirement for correctness.

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225 chauhang

[ghstack-poisoned]

Commit:
Update on "[FSDP2] Added 2D DCP save/load test"


To prepare for FSDP2 + TP/SP in torchtrain, we should verify that we can resume training correctly with DCP save/load. For loading into a new model/optimizer instance, torchtrain uses lightweight `ModelWrapper` and `OptimizerWrapper`. In the added unit test, we use `get_optimizer_state_dict` directly to show the minimal requirement for correctness.

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225 chauhang

[ghstack-poisoned]

Commit:
Update on "[FSDP2] Added 2D DCP save/load test"


To prepare for FSDP2 + TP/SP in torchtrain, we should verify that we can resume training correctly with DCP save/load. For loading into a new model/optimizer instance, torchtrain uses lightweight `ModelWrapper` and `OptimizerWrapper`. In the added unit test, we use `get_optimizer_state_dict` directly to show the minimal requirement for correctness.

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225 chauhang

[ghstack-poisoned]

Commit:
Update on "[FSDP2] Added 2D DCP save/load test"


To prepare for FSDP2 + TP/SP in torchtrain, we should verify that we can resume training correctly with DCP save/load. For loading into a new model/optimizer instance, torchtrain uses lightweight `ModelWrapper` and `OptimizerWrapper`. In the added unit test, we use `get_optimizer_state_dict` directly to show the minimal requirement for correctness.

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225 chauhang

[ghstack-poisoned]

wz337 commented on test/distributed/_composable/fsdp/test_fully_shard_training.py:
```diff
@@ -710,13 +713,16 @@ class TestFullyShard2DTraining(FSDPTest):
         )
+
+    @skip_if_lt_x_gpu(2)
+    def test_train_parity_2d_mlp(self):
```
Should we make optimizer an option in .run_subtests? 
Actually, I can try this after you land, since I have observed parameters are off by a little bit more than what torch.allclose() tolerates by default in the unit tests.

Context: https://github.com/pytorch/pytorch/pull/121774
Local test is fine but CI is failing.

awgu replied:
I think since Adam is fine/passing, maybe we can leave the test always testing optimizer here since we want to check the end-to-end resume flow? Then, you can investigate the AdamW case separately?

wz337 commented on test/distributed/_composable/fsdp/test_fully_shard_training.py:
```diff
@@ -782,6 +788,106 @@ def _test_train_parity_2d_mlp(
+        self.assertEqual(loss_no_cp1, loss_cp1)
+
+        sharded_sd = {
+            "model": model_cp.state_dict(),
```
For the sake of testing DCP checkpoint, could we also use `get_model_state_dict()`, as we expect `model.state_dict()` to be identical as `get_model_state_dict()`?

wz337 reviewed (approved):
LGTM!

The comments are minor. If it is easier for you, you can land it as it is. I can try AdamW and update model state dict to use distributed state dict as well.

Commit:
Update on "[FSDP2] Added 2D DCP save/load test"


To prepare for FSDP2 + TP/SP in torchtrain, we should verify that we can resume training correctly with DCP save/load. For loading into a new model/optimizer instance, torchtrain uses lightweight `ModelWrapper` and `OptimizerWrapper`. In the added unit test, we use `get_optimizer_state_dict` directly to show the minimal requirement for correctness.

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225 chauhang

[ghstack-poisoned]

Commit:
Update on "[FSDP2] Added 2D DCP save/load test"


To prepare for FSDP2 + TP/SP in torchtrain, we should verify that we can resume training correctly with DCP save/load. For loading into a new model/optimizer instance, torchtrain uses lightweight `ModelWrapper` and `OptimizerWrapper`. In the added unit test, we use `get_optimizer_state_dict` directly to show the minimal requirement for correctness.

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225 chauhang

[ghstack-poisoned]

awgu commented:
@pytorchbot merge
